decision-records/2024-08-29--dashboard-collision.md

```diff
@@ -0,0 +1,144 @@
+# Dashboard collisions in Grafana
+**Date:** 2024-08-29<br/>
+**Authors:** @lucabello
+
+
+## Context and Problem Statement
+Dashboard json includes keys such as in the following example:
+```bash
+$ cat metrics-dashboard.json | jq '{"title", "uid", "version", "tags"}'
+{
+  "title": "Loki Operator Overview",
+  "uid": "loki",
+  "version": 1,
+  "tags": [
+    "loki"
+  ]
+}
+```
+
+Grafana [documented](https://grafana.com/docs/grafana/latest/administration/provisioning/#reusable-dashboard-urls) that dashboards should not have the same *title* within a folder, or the same *uid* within installations.
+
+This causes some problems in different scenarios.
+
+### Problem 1: UID collision
+
+#### (a) Two charms on different revisions
+
+`charm(rev1, dashboard v1) <--> grafana <--> charm(rev2, dashboard v2)`
+
+Unless manually modified, `dashboard v1` and `dashboard v2` have the same *title* and *uid*, which will cause problems in Grafana.
+
+#### (b) New dashboard created by cloning an existing dashboard, neglecting to modify the UID
+
+A common way to create a new dashboard is to clone and modify dashboard (local or from the grafana marketplace).
+If the UID isn't modified (need to remeber to do it manually), then we'll end up with a UID collision (but the revision would be auto-incremented by the grafana UI, if saved and exported using grafana).
+
+#### (c) Unrelated dashboards that coincidentally happen to have the same UID
+
+This is likely to happen when charm authors manually modify UIDs, assuming their (naive :) modification produced a unique ID (in the ecosystem!).
+
+### Problem 2: Title collision
+
+Dashboards from different charms may happen to have the same title:
+
+`loki(dashboard "Overview") <--> grafana <--> tempo(dashboard "Overview")`
+
+If both charms have a dashboard simply named "Overview", we have problems. Suggesting a UID and Title to be manually upgraded on every dashboard modification (to avoid (1)) seems like a tedious UX.
+
+## Requirements
+1. Dashboards must not have the same *uid* within installation.
+2. Dashboards must not have the same *title* within a folder.
+3. Dashboard URL must be stable across dashboard updates (e.g. charm upgrades). Do not generate the UID from the hasing dashboard's entire contents, because it would break links to the dashboard.[^DbURL]
```

Comment:
I think this is a typo and you mean "... by hashing the dashboard's contents, ..."